src/gt4py/next/type_system/type_info.py

```diff
@@ -681,36 +680,47 @@ def canonicalize_function_arguments(
     kwargs: dict,
     *,
     ignore_errors: bool = False,
-    use_signature_ordering: bool = False,
 ) -> tuple[tuple, dict]:
-    num_pos_params = len(func_type.pos_only_args) + len(func_type.pos_or_kw_args)
-    cargs = [UNDEFINED_ARG] * max(num_pos_params, len(args))
-    ckwargs = {**kwargs}
-    for i, arg in enumerate(args):
-        cargs[i] = arg
-    for name in kwargs.keys():
+    num_pos_only_args = len(func_type.pos_only_args)
+    num_pos_or_kw_args = len(func_type.pos_or_kw_args)
+    canonical_args = [*args] + (
+        [UNDEFINED_ARG] * max(num_pos_only_args + num_pos_or_kw_args - len(args), 0)
+    )
+    remaining_kwargs = {**kwargs}
+
+    pos_or_kw_args_names = [*func_type.pos_or_kw_args]
+    for name in kwargs:
         if name in func_type.pos_or_kw_args:
-            args_idx = len(func_type.pos_only_args) + list(func_type.pos_or_kw_args.keys()).index(
-                name
-            )
-            if cargs[args_idx] is UNDEFINED_ARG:
-                cargs[args_idx] = ckwargs.pop(name)
+            args_idx = num_pos_only_args + pos_or_kw_args_names.index(name)
+            if canonical_args[args_idx] is UNDEFINED_ARG:
+                canonical_args[args_idx] = remaining_kwargs.pop(name)
             elif not ignore_errors:
-                raise AssertionError(
+                raise ValueError(
                     f"Error canonicalizing function arguments. Got multiple values for argument '{name}'."
                 )
 
-    a, b = set(func_type.kw_only_args.keys()), set(ckwargs.keys())
-    invalid_kw_args = (a - b) | (b - a)
-    if invalid_kw_args and (not ignore_errors or use_signature_ordering):
-        # this error can not be ignored as otherwise the invariant that no arguments are dropped
-        # is invalidated.
-        raise AssertionError(f"Invalid keyword arguments '{', '.join(invalid_kw_args)}'.")
+    missing_kw_args = func_type.kw_only_args.keys() - remaining_kwargs.keys()
+    invalid_kw_args = remaining_kwargs.keys() - func_type.kw_only_args.keys()
+
+    if not ignore_errors:
+        if missing_kw_args:
+            raise ValueError(f"Missing required keyword arguments: {[*missing_kw_args]}.")
+        if invalid_kw_args:
+            raise ValueError(f"Invalid keyword arguments: {[*invalid_kw_args]}.")
 
-    if use_signature_ordering:
-        ckwargs = {k: ckwargs[k] for k in func_type.kw_only_args.keys() if k in ckwargs}
+        # Sort remaining keyword arguments in the signature ordering
+        canonical_kwargs = {
+            k: remaining_kwargs[k] for k in func_type.kw_only_args if k in remaining_kwargs
+        }
```

Comment:
```suggestion
        canonical_kwargs = {
            k: remaining_kwargs[k] for k in func_type.kw_only_args if k in remaining_kwargs
        } | {remaining_kwargs[k] for k in invalid_kw_args}
```
Then the else can be removed below.